Extension/src/LanguageServer/client.ts

```diff
@@ -712,6 +712,7 @@ export class DefaultClient implements Client {
                                         this.client.languageClient.sendNotification(RenameNotification, params);
                                         this.client.references.setResultsCallback((final, referencesResult) => {
                                             referencesRequestPending = false;
+                                            renamePending = false;
```

Comment:
>renamePending = false; [](start = 44, length = 22)

Could you remove this one?  It should be cleared as soon as cancellation is requested.  If this one is actually needed, let me know.  That might point to another problem.

Reply:
This one is needed, otherwise it will issue a cancel after a "commit rename" is completed. The cancel will get issues from `onDidChangeTextDocument(...)` because `renamePending` is true.

Reply:
Actually, this one is a problem.  It will clear renamePending, but it's possible for another rename to be queued after this that has already set renamePending to true, and this will stomp on it.

Reply:
Prior to removing the check for whether an request is really in progress in cancelReferences(), this was not an issue.  I think we need that check restored, but also check for the state in which a preview occurred instead, rather than always issue that cancallation.

Reply:
(In response to my deleted comment you replied to).  I think my above 2 responses cover the issue.

Reply:
> but it's possible for another rename to be queued after this that has already set renamePending to true, and this will stomp on it.

this `renamePending` does get stomped on it when an edit happens during rename pending state.

> I think we need that check restored, but also check for the state in which a preview occurred instead, rather than always issue that cancallation.

Will try checking for a "preview occurred"

Reply:
I was referring to it being set to true on line 695.  By seeing it to false against on line 715, you are stomping on a subsequent Rename operation (not the one that was cancelled).  renamePending gets set outside of the 'queue' of pending operations.  As long as there is any rename pending, it should be true, and should only be cleared when a cancel occurs (as any pending rename should be cancelled by that).

Reply:
Yep, that's what I ran into when I debugged the scenario:

1. do a rename
2. while a rename is pending, send another rename request

Reply:
I think there is still a problem here, even if we restore the check that was removed from cancelReferences.  We need to ensure that pendingRename gets cleared after the operation is completed successfully (only if there are no other renames pending), or onDidChangeTextDocument might wind up cancelling the next FAR operation thinking it's a rename.  ...  

---
In reply to: [328849839](https://github.com/Microsoft/vscode-cpptools/pull/4342#discussion_r328849839) [](ancestors = 328849839)

Reply:
I think we could address this with a counter that is incremented at the very start of each rename, and decremented on each path out of rename (whenever rejected or completed).  If decremented to 0, clear renamePending.   I could create a separate PR to do these things I'm suggesting, if you would prefer.

---
In reply to: [328870043](https://github.com/Microsoft/vscode-cpptools/pull/4342#discussion_r328870043) [](ancestors = 328870043,328849839)